src/win/pipe.c

```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
 }
 
 
+static DWORD WINAPI uv__readfile_interrupter_thread(void* param) {
+  uv__readfile_interrupter_t* interrupter;
+  int r;
+
+  interrupter = param;
+  for (;;) {
+    r = WaitForSingleObject(interrupter->completed_event,
+                            PIPE_ZERO_READ_INTERRUPT_INTERVAL);
+    if (r == WAIT_TIMEOUT) {
+      /* ReadFile thread is active - interrupt ReadFile to give other */
+      /* processes a chance to act upon pipe */
+      WaitForSingleObject(interrupter->thread_mutex, INFINITE);
+      if (interrupter->thread != NULL) {
+        pCancelSynchronousIo(interrupter->thread);
+        SwitchToThread();
+      }
+      ReleaseMutex(interrupter->thread_mutex);
+    } else if (r == WAIT_OBJECT_0) {
+      /* ReadFile thread has terminated - clean up the handles and exit */
+      CloseHandle(interrupter->completed_event);
+      CloseHandle(interrupter->thread_mutex);
+      uv__free(interrupter);
+      return 0;
+    }
+  }
+}
+
+static uv__readfile_interrupter_t* uv__start_readfile_interrupter(void) {
+  HANDLE thread_handle;
+  uv__readfile_interrupter_t* interrupter;
+
+  interrupter = uv__malloc(sizeof(*interrupter));
+  if (interrupter == NULL)
+    return NULL;
+
+  if (!DuplicateHandle(GetCurrentProcess(),
+                       GetCurrentThread(),
+                       GetCurrentProcess(),
+                       &thread_handle,
+                       0,
+                       FALSE,
+                       DUPLICATE_SAME_ACCESS))
+    goto failed_duplicate;
+
+  interrupter->thread = thread_handle;
+  interrupter->thread_mutex = CreateMutex(NULL, FALSE, NULL);
+  if (interrupter->thread_mutex == NULL)
+    goto failed_mutex;
+
+  interrupter->completed_event = CreateEvent(NULL, FALSE, FALSE, NULL);
+  if (interrupter->completed_event == NULL)
+    goto failed_event;
+
+  if (!QueueUserWorkItem(uv__readfile_interrupter_thread,
+                         interrupter,
+                         WT_EXECUTEINLONGTHREAD))
```

Comment:
[MSDN](https://msdn.microsoft.com/pl-pl/library/windows/desktop/ms685061.aspx) says I can't wait on a pipe.

I could use [CreateTimerQueueTimer](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682485.aspx) to use the threadpool to wake interrupter every 2.5s, but I think that would complicate things.